src/vc.rs

```diff
@@ -537,29 +592,229 @@ impl Credential {
+                    VerificationResult::error(&format!("Unable to decode JWS: {}", err)),
+                );
+            }
+        };
+        let claims: JWTClaims = match serde_json::from_slice(&payload) {
+            Ok(claims) => claims,
+            Err(err) => {
+                return (
+                    None,
+                    VerificationResult::error(&format!("Unable to decode JWS claims: {}", err)),
+                );
+            }
+        };
+        let vc = match Self::from_jwt_claims(claims.clone()) {
+            Ok(claims) => claims,
+            Err(err) => {
+                return (
+                    None,
+                    VerificationResult::error(&format!(
+                        "Unable to convert JWT claims to VC: {}",
+                        err
+                    )),
+                );
+            }
+        };
+        if let Err(err) = vc.validate_unsigned() {
+            return (
+                None,
+                VerificationResult::error(&format!("Invalid VC: {}", err)),
+            );
+        }
+        // TODO: error if any unconvertable claims
+        // TODO: unify with verify function?
+        let (proofs, matched_jwt) = match vc
+            .filter_proofs(options_opt, Some((&header, &claims)), resolver)
+            .await
+        {
+            Ok(matches) => matches,
+            Err(err) => {
+                return (
+                    None,
+                    VerificationResult::error(&format!("Unable to filter proofs: {}", err)),
+                );
+            }
+        };
+        let verification_method = match header.key_id {
+            Some(kid) => kid,
+            None => {
+                return (
+                    None,
+                    VerificationResult::error(&format!("JWT header missing key id")),
+                );
+            }
+        };
+        let key = match crate::ldp::resolve_key(&verification_method, resolver).await {
+            Ok(key) => key,
+            Err(err) => {
+                return (
+                    None,
+                    VerificationResult::error(&format!("Unable to resolve key for JWS: {}", err)),
+                );
+            }
+        };
+        let mut results = VerificationResult::new();
+        if matched_jwt {
+            match crate::jws::verify_bytes(header.algorithm, &signing_input, &key, &signature) {
+                Ok(()) => results.checks.push(Check::JWS),
+                Err(err) => results
+                    .errors
+                    .push(format!("Unable to filter proofs: {}", err)),
+            }
+            return (Some(vc), results);
+        }
+        // No JWS verified: try to verify a proof.
+        if proofs.is_empty() {
+            return (
+                None,
+                VerificationResult::error("No applicable JWS or proof"),
+            );
+        }
+        // Try verifying each proof until one succeeds
+        for proof in proofs {
```

Comment:
Could the proofs be filtered first to make it faster?

Reply:
They are filtered earlier in the function.

Reply:
errrrrr I'm an idiot